src/Futhark/CodeGen/Backends/GenericC.hs

```diff
@@ -677,6 +693,40 @@ prepareArg (MemParam name (Space sid)) = do
 
 prepareArg p = return [C.cexp|$exp:(paramName p)|]
 
+prepareMemCopy :: Param -> CompilerM op s (Maybe VName)
+prepareMemCopy (MemParam name space) = do
+  name_copy <- newVName $ baseString name <> "_copy"
+  declMem name_copy space
+  let size = [C.cexp|$id:name.size|]
+
+  alloc_mem <- collect $ allocMem name_copy size space
+  item [C.citem|
+        if (perform_warmup || num_runs > 1) {
+          $items:alloc_mem
+        }
+       |]
+  return $ Just name_copy
+prepareMemCopy _ = return Nothing
+
+copyMemBeforeRun :: Param -> C.Exp -> Maybe VName -> CompilerM op s C.Exp
+copyMemBeforeRun (MemParam name space) arg (Just arg_mem_copy) = do
+  copy <- asks envCopy
+  let size = [C.cexp|$id:name.size|]
+      dest = rawMem' True arg_mem_copy
+      src = rawMem' True arg
+  case space of
+    DefaultSpace ->
+      stm [C.cstm|memmove($exp:dest, $exp:src, $exp:size);|]
+    _ ->
+      copy dest [C.cexp|0|] space src [C.cexp|0|] space size
```

Comment:
You should add a universal copy function that uses opsCopy if necessary. It's amazing we don't already have that.